File: 0x00 基本算法/0x06 倍增/AcWing109.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long LL;
const int N = 5e5 + 10;
int n, m, k, a[N], b[N], tmp[N]; // 必须引入 b，用于存储目前已经确立的、排好序的左半部分区间
LL t;

// 检查如果把区间从 mid 扩展到 r，新的这段区间 [l, r] 是否合法
bool check(int l, int mid, int r) {
  for (int i = mid + 1; i <= r; i ++ ) b[i] = a[i];
  sort(b + mid + 1, b + r + 1); // 将新步长的一截 [mid+1, r] 提取到 b 数组并单独排序

  int i = l, j = mid + 1; // 将 b[l~mid] 和 b[mid+1~r] 合并到 tmp
  for (int k = l; k <= r; k ++ )
    if (j > r || (i <= mid && b[i] <= b[j])) tmp[k] = b[i ++ ];
    else tmp[k] = b[j ++ ];

  LL sum = 0; // 贪心计算最大校验差的平方和
  for (int i = l, j = r; i < min(l + m, j); i ++ , j -- )
    sum += 1LL * (tmp[i] - tmp[j]) * (tmp[i] - tmp[j]);
  return sum <= t; // 这里绝对不能把 tmp 抄回 b，因为不知道是否合法
}

int main() {
  cin >> k;
  while (k -- ) {
    cin >> n >> m >> t;
    for (int i = 0; i < n; i ++ ) {
      cin >> a[i];
      b[i] = a[i];
    }

    int cnt = 0, l = 0;
    while (l < n) {
      int p = 1, r = l;
      while (p)
        if (r + p < n && check(l, r, r + p)) { // 试探将区间向右扩展 p 的长度
          r += p;
          for (int i = l; i <= r; i ++ ) b[i] = tmp[i]; // 成功，此时把合法的 tmp 更新回 b
          p <<= 1;
        }
        else p >>= 1; // 试探失败或越界，步长减半

      l = r + 1;
      cnt ++ ;
    }
    cout << cnt << endl;
  }
  return 0;
}

File: 0x00 基本算法/0x06 倍增/AcWing109.py
import sys

def get_ints():
    for line in sys.stdin:
        for token in line.split():
            yield int(token)

def main():
    iterator = get_ints()
    
    try:
        k = next(iterator)
    except StopIteration:
        return
        
    for _ in range(k):
        n = next(iterator)
        m = next(iterator)
        t = next(iterator)
        
        a = [next(iterator) for _ in range(n)]
        b = a[:]
        tmp = [0] * n
        
        def check(l, mid, r):
            b[mid + 1:r + 1] = sorted(a[mid + 1:r + 1])
            
            i, j = l, mid + 1
            for idx in range(l, r + 1):
                if j > r or (i <= mid and b[i] <= b[j]):
                    tmp[idx] = b[i]
                    i += 1
                else:
                    tmp[idx] = b[j]
                    j += 1
                    
            total_sum = 0
            i, j = l, r
            pairs = 0
            while i < j and pairs < m:
                diff = tmp[i] - tmp[j]
                total_sum += diff * diff
                if total_sum > t:
                    return False
                i += 1
                j -= 1
                pairs += 1
                
            return True

        cnt = 0
        l = 0
        while l < n:
            p = 1
            r = l
            while p > 0:
                if r + p < n and check(l, r, r + p):
                    r += p
                    b[l:r + 1] = tmp[l:r + 1]
                    p <<= 1
                else:
                    p >>= 1
            l = r + 1
            cnt += 1
            
        print(cnt)

if __name__ == '__main__':
    main()